client.go

```diff
@@ -0,0 +1,143 @@
+	CompletionStream(ctx context.Context, request CompletionRequest, onData func(*CompletionResponse)) error
+	Edits(ctx context.Context, request EditsRequest) (*EditsResponse, error)
+	Embeddings(ctx context.Context, request EmbeddingsRequest) (*EmbeddingsResponse, error)
+	Files(ctx context.Context) (*FilesResponse, error)
+	UploadFile(ctx context.Context, request UploadFileRequest) (*FileObject, error)
+	DeleteFile(ctx context.Context, fileID string) (*DeleteFileResponse, error)
+	File(ctx context.Context, fileID string) (*FileObject, error)
+	FileContent(ctx context.Context, fileID string) ([]byte, error)
+	CreateFineTune(ctx context.Context, request CreateFineTuneRequest) (*FineTuneObject, error)
+	ListFineTunes(ctx context.Context) (*ListFineTunesResponse, error)
+	FineTune(ctx context.Context, fineTuneID string) (*FineTuneObject, error)
+	CancelFineTune(ctx context.Context, fineTuneID string) (*FineTuneObject, error)
+	FineTuneEvents(ctx context.Context, request FineTuneEventsRequest) (*FineTuneEventsResponse, error)
+	FineTuneStreamEvents(ctx context.Context, request FineTuneEventsRequest, onData func(*FineTuneEvent)) error
+	DeleteFineTuneModel(ctx context.Context, modelID string) (*DeleteFineTuneModelResponse, error)
+
+	// Deprecated
+	CompletionWithEngine(ctx context.Context, engine string, request CompletionRequest) (*CompletionResponse, error)
+	CompletionStreamWithEngine(ctx context.Context, engine string, request CompletionRequest, onData func(*CompletionResponse)) error
+}
+
+type client struct {
+	baseURL      string
+	apiKey       string
+	orgID        string
+	userAgent    string
+	httpClient   *http.Client
+	defaultModel string
+}
+
+func NewClient(apiKey string, options ...ClientOption) (Client, error) {
+	c := &client{
+		baseURL:      DEFAULT_BASE_URL,
+		apiKey:       apiKey,
+		orgID:        "",
+		userAgent:    DEFAULT_USER_AGENT,
+		httpClient:   &http.Client{Timeout: time.Duration(DEFAULT_TIMEOUT) * time.Second},
+		defaultModel: DavinciModel,
+	}
+
+	for _, option := range options {
+		if err := option(c); err != nil {
+			return nil, err
+		}
+	}
+
+	return c, nil
+}
+
+func (c *client) newRequest(ctx context.Context, method, path string, payload interface{}) (*http.Request, error) {
+	bodyReader, err := jsonBodyReader(payload)
+	if err != nil {
+		return nil, err
+	}
+	url := c.baseURL + path
+	req, err := http.NewRequestWithContext(ctx, method, url, bodyReader)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-type", "application/json")
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.apiKey))
+	req.Header.Set("User-Agent", c.userAgent)
+	if len(c.orgID) > 0 {
+		req.Header.Set("OpenAI-Organization", c.orgID)
+	}
+	return req, nil
+}
+
+func (c *client) performRequest(req *http.Request) (*http.Response, error) {
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
```

Comment:
I was testing out the changes on this branch locally and was running into errors, that I believe are from this line. I don't think we used to do this defer/body close from the `performRequest` happy path, so it was giving me an error when trying to run `go run cmd/test/main.go`
> invalid json response: http2: response body closed

Reply:
Ah let me fix that, I believe I was told to move it there, does this occur when there is no response body?